managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/MetaStoreImpl.java

```diff
@@ -453,8 +456,13 @@ public ManagedLedgerInfo parseManagedLedgerInfo(byte[] data) throws InvalidProto
             try {
                 MLDataFormats.ManagedLedgerInfoMetadata metadata =
                         MLDataFormats.ManagedLedgerInfoMetadata.parseFrom(metadataBytes);
-                return ManagedLedgerInfo.parseFrom(getCompressionCodec(metadata.getCompressionType())
-                        .decode(byteBuf, metadata.getUncompressedSize()).nioBuffer());
+                ByteBuf decode = getCompressionCodec(metadata.getCompressionType())
+                        .decode(byteBuf, metadata.getUncompressedSize());
+                try {
+                    return ManagedLedgerInfo.parseFrom(decode.nioBuffer());
+                } finally {
+                    decode.release();
```

Comment:
When using LightProto, the buffer shouldn't be released until the parsed instance is no longer used. This is due to the implementation detail that String properties will be lazily resolved from the input buffer.
It's possible to resolve this by making a copy with `ManagedLedgerInfo.copyFrom(ManagedLedgerInfo.parseFrom(decode.nioBuffer())` unless there's a solution in place to ensure that buffers aren't released as long as the instance is used.

Reply:
Maybe I am missing something, but as I understand:

* parseFrom() needs a nio buffer.
* .nioBuffer() will return nio buffer with shared data if ByteBuff is backed by heap array/nio buffer, otherwise it will duplicate the buffer into a nio buffer.
* ManagedLedgerInfo does not have .release() method or equivalent
* ManagedLedgerInfo does not have .copyFrom() method

decode not being released was a bug that can go unnoticed in some cases, or result in e.g. slow direct memory leak in others.

I guess you were thinking about google's protobuf with perf optimized/manually constructed CodedInputStream/ByteString.

Reply:
This is fine, `ManagedLedgerInfo` is generated by Google Protobuf's protoc. I thought it was created with LightProto. In that case my comment would have applied.